### ISSUE TITLE
Widget builder methods should take `&mut self` instead of `mut self`

### ISSUE THREAD

**Issue by ruza-net** [iced-rs/iced#697]
Not only is it a good practice, it's also beneficial in situations where you don't have ownership of a widget but still want to alter its properties.

Use case:
```rust
// Loads an image together with some metadata, owns the corresponding `Image`
//
pub struct Icon { /* ... */ }

// Owns an `Icon` and on `draw` wraps it in a nice border
//
pub struct ImageButton { /* ... */ }

impl ImageButton {
    pub fn width(&mut self, len: Length) -> &mut Self {
        self.icon.image_mut().width(len); // Today, doesn't compile, since cannot move out of mutable reference.
        self
    }
}
```

**Comment by twitchyliquid64** [comment]
Given you have mutable control of the icon field,  couldn't you do this?

```rust
self.icon = self.icon.width(len);
```

> Not only is it a good practice

Could you link me to where this is from? Not arguing or anything, but as I'm self-taught I'm always interested in reading an authoritative list of best practices.

**Comment by ruza-net** [comment]
> Given you have mutable control of the icon field, couldn't you do this?

Well, in this example, yes I suppose. However, consider a method which has to `view` or `update` the component. These typically take a mutable reference, because the caller also has a `&mut self` (because `Application` requires that). Ultimately at some level, someone has to [`replace_with`](https://crates.io/crates/replace_with).

> Could you link me to where this is from? Not arguing or anything, but as I'm self-taught I'm always interested in reading an authoritative list of best practices.

This is a Reddit discussion referencing a post by Aaron Turon (one of Rust's internal developers), which unfortunately doesn't exist anymore:
[Builder Pattern &mut self vs mut self](https://www.reddit.com/r/rust/comments/5fx2pt/builder_pattern_mut_self_vs_mut_self/)

The tl;dr of the discussion is that `&mut self` has the benefit of caching and easier copying, whereas the benefit of `mut self` is in state machines (you can enforce certain order of builder methods by returning a newtype from them).

I've changed my model since when I posted this, and I don't need the `&mut self` behavior anymore. I suppose it's more ergonomic to write:

```rust
Button::new().on_press(..).style(..).into()
```

vs:

```rust
let btn = Button::new();

btn.on_press(..).style(..);

btn.into()
```

I think there should maybe be an API for both.

**Comment by hecrj** [comment]
We use `mut self` because defaulting to immutability seems to suit Elm's approach best. I personally like to keep an eye on the amount of `let mut` in my code.

Additionally, while we are not returning new types from the widget methods as of now, it's something we may rely on eventually and having an inconsistent style across widgets would be very confusing.
